lib/mongo_client.js

```diff
@@ -206,6 +207,32 @@ class RemoteMongoDBCollection {
             upsert: options.upsert,
         });
     }
+
+    async* watch({ids = undefined, filter = undefined} = {}) {
```

Comment:
This is the first generator function in Realm JS. Is there a different way of achieving this without introducing the use of a new ES feature to the users? I'm thinking ex the observer pattern (event emitters) are used extensively throughout the API.

Reply:
I was specifically *avoiding* using an event emitter-style API here. They have the significant problem with all push-based APIs that the consumer has no way of controlling when they are ready for the next event, and therefore no way of providing backpressure. Using async iterables/iterators (whether via a generator or directly implementing the `@@asyncIterator` protocol) uses a pull-style API that puts the client in control of when they want the next event. This lets you do things like this, and ensure that events are processed in order:

```js
for await (let event of collection.watch()) {
    await notifyServerOfEvent(event);
}
```

That is not possible to implement correctly if we were using the event emitter pattern. This is a general case of pull-style iteration being more flexible and generally preferable to push-style. Note that it is trivial to wrap a pull-style iterator into a push-style one, but impossible to go back. That means that if we only provided a emitter API, developers *wouldn't be able* to use `for await`-style iteration. If they want to use event emitters, this is all it takes:

```js
function onWatchEvent(watch, handler) {
  (async() => {
      for await (let event of watch)) handler(event);
  )().onError(e => handler(undefined, e));
}```

We could of course offer that helper as a separate function, but I'm not sure when you would ever prefer to use that rather than just directly consuming the async iterator.